### DISCUSSION TITLE
"Proxy" option similar to redirects

### DISCUSSION THREAD

**Discussion by japem** [vercel/next.js#43499] [Ideas]
### Describe the feature you'd like to request

The Next.js configuration currently provides a `redirects` option which issues 3xx responses to API requests to redirect them to other routes. This is a problem for use cases where the API handlers are not written in Next.js when doing local development. I cannot use `redirects` to tell my application to redirect the API request to some other API server - be it one running locally on a different port or on some remote nonprod server - because my browser may refuse to follow a redirect to a different host.

If Next.js provided a `proxies` option, which worked similarly but did the proxying itself rather than issuing a 3xx response, this would solve the problem, and allow me to test my app locally while making API calls to some other server.

This doesn't even have to work when Next.js is not running development mode, as I imagine it could lead to all sorts of issues where we actually care about CORS. I also imagine this is less of an issue in non-local environments, where reverse proxies at the edge can be used instead of redirects to divert API requests to other locations, or API servers are hosted on the same domain as the Next.js app so redirects may be used.

### Describe the solution you'd like

An option `proxies` in `next.config.js`, sibling to `redirects`, which is similarly a function that returns an array of objects each configuring different proxies. The exact specification of these objects isn't particularly important, but it could be similar to that of `redirects` or `rewrites`.

A simple implementation could use [http-proxy](https://github.com/http-party/node-http-proxy). The specification of the configuration could closely match the `http-proxy` options, similarly to [how Vite accomplishes this](https://vitejs.dev/config/server-options.html#server-proxy).

### Describe alternatives you've considered

A simple workaround to this problem is to create my own API routes that do this proxying. This provides a good amount of flexibility around how this proxying happens, modifications to make to the request, etc. However, this is cumbersome at best a security risk at worst. A configuration option that did all this work for me would make things significantly simpler, especially in the case where I have different hosts for different APIs, constantly evolving host addresses, etc; it's also not ideal to have this API-to-host mapping exist in code rather than in configuration.

More importantly, these API routes will exist in my production app where I don't want them to be used - I can try to hide them with redirects, but that's a messy solution.

Configuring a browser to force it to follow all redirects may be possible but is clearly not a viable solution.

**Comment by raihle** [comment]
How is this different from `rewrites`?

**Reply by japem** [reply]
Rewrites are applied to client-side routing. I want this for API routes.

For example, if I have an API separate from my Next.js app served at localhost:8000/api, I cannot use rewrites to cause `fetch("/api")` to be proxied to localhost:8000/api. I must call `fetch("http://localhost:8000/api")`, which I likely do not want to be in my production code.
